### ISSUE TITLE
Problem with running/authenticating on an EC2 (AWS) machine and checking right cert/key files used

### ISSUE THREAD

**Issue by Analect** [jupyterhub/jupyterhub-deploy-docker#6]
@jtyberg 
Thanks for dockerizing this. I followed your detailed README.
The only part that wasn't clear was what to use for jupyterhub.cer / jupyterhub.key files in the secret folder. I assumed from looking at the compose file [here](https://github.com/jupyterhub/jupyterhub-deploy-docker/blob/11fddae0a79bdf5babba7f63d192ad6fda07f521/docker-compose.yml#L29-L30) that I should be using the `server.pem` and `server-key.pem` files from the `/.docker/machine/machines/jupyterhub/` folder (as per below).  Is that the case? And should I copy these over to the a `secrets` folder in this repo as `jupyterhub.cer` and `jupyterhub.key` respectively?

```
-rw-------  1 me staff   381  4 May 15:35 id_rsa.pub
-rw-------  1 me staff  1679  4 May 15:35 id_rsa
-rw-r--r--  1 me staff  1127  4 May 15:39 server.pem
-rw-------  1 me staff  1679  4 May 15:39 server-key.pem
-rw-------  1 me staff  1675  4 May 15:39 key.pem
-rw-r--r--  1 me staff  1078  4 May 15:39 cert.pem
-rw-r--r--  1 me staff  1038  4 May 15:39 ca.pem
-rw-------  1 me staff  3407  4 May 15:39 config.json
```
On running up the jupyterhub container, I get this in the logs, which looks OK, I think.
```
jupyterhub | [I 2016-05-04 16:04:01.700 JupyterHub app:595] Loading cookie_secret from /data/jupyterhub_cookie_secret
jupyterhub | [W 2016-05-04 16:04:01.734 JupyterHub app:300]
jupyterhub |     Generating CONFIGPROXY_AUTH_TOKEN. Restarting the Hub will require restarting the proxy.
jupyterhub |     Set CONFIGPROXY_AUTH_TOKEN env or JupyterHub.proxy_auth_token config to avoid this message.
jupyterhub |
jupyterhub | [I 2016-05-04 16:04:01.764 JupyterHub app:1155] Hub API listening on http://jupyterhub:8080/hub/
jupyterhub | [I 2016-05-04 16:04:01.767 JupyterHub app:902] Starting proxy @ http://*:443/
jupyterhub | 16:04:01.865 - info: [ConfigProxy] Proxying https://*:443 to http://jupyterhub:8080
jupyterhub | 16:04:01.867 - info: [ConfigProxy] Proxy API at http://127.0.0.1:444/api/routes
jupyterhub | [I 2016-05-04 16:04:01.872 JupyterHub app:1178] JupyterHub is now running at http://127.0.0.1:443/
```
Navigating to the host in the browser, I'm presented with the Github login button, and this initially appears to work (seeking me to authenticate in Github against the application I had created).  But it somehow doesn't complete that handshake .. and ends up giving me a 403 and then a 500 server error (as per the screenshots below).

![image](https://cloud.githubusercontent.com/assets/4063815/15027516/229ea860-123c-11e6-9d71-0453b217235d.png)

Here are the logs from the jupyterhub container, which are probably more helpful in terms of figuring out what might be going wrong: https://gist.github.com/Analect/695752f105441d250517cfadcef6cc6a

Just to note that I have ports `8000`, `2376` `444` and `443` open on the EC2 instance, so it shouldn't really be a security blocking problem.

Would you have any pointers for me as to what might be going wrong here?  Thanks.

**Comment by jtyberg** [comment]
The JupyterHub certificate and key files are for the JupyterHub server to support HTTPS.  They should be generated for your domain (https://jupyterhub.my.domain).  You can either use self-signed certs, or something like Let's Encrypt to generate real ones (see `examples/letsencrypt`).   The certs in the `/.docker/machine/machines/jupyterhub` directory are the TLS certs that docker-machine uses to authenticate to the Docker daemon on the host, so definitely don't use those.

Once you put proper certificate and key files in place, make sure you registered the OAuth callback properly on GitHub.  It should be in the format: https://jupyterhub.my.domain/hub/oauth_callback.

Let us know if you are still experiencing problems after fixing the certs and double-checking the callback URL.

**Comment by Analect** [comment]
@jtyberg Thanks for the quick response ... and for clarifying my confusion around the certs. I have tried to follow your example/letsencrypt implementation, but for some reason it keeps failing on connecting to verify the domain, and therefore won't generate a certificate, eventhough I do have port 443 open on that EC2 instance.

I note from this [issue](https://github.com/letsencrypt/letsencrypt/issues/2462), someone having similar issues, in that he's using a container within an EC2 instance to set up the encryption keys ... so I tried modifying your letsencrypt.sh file [here](https://github.com/jupyterhub/jupyterhub-deploy-docker/blob/master/examples/letsencrypt/letsencrypt.sh#L61) to expose `p 443:443` from the container, to see if that resolved it, but it didn't.

The A record for this domain exists and points to the IP of the EC2 instance.  I have obfuscated the name of my domain below.  And, as I said, port 443 is open on that instance, so there isn't anything obvious to me that could be blocking that connection call to verify.

In the meantime, I'll try the self-signed option, just to get things up and running.

```
me$ ./letsencrypt.sh --domain jupyterhub.xxxxx.com  --email me@mydomain.com --volume jupyterhub-secrets
Failed authorization procedure. jupyterhub.xxxxx.com (http-01): urn:acme:error:connection :: The server could not connect to the client to verify the domain :: Could not connect to http://jupyterhub.xxxxx.com/.well-known/acme-challenge/s7Amlt5iyVBMhANWEizToY7-PdveDtKNgHBVIaIeyP0

IMPORTANT NOTES:
 - The following errors were reported by the server:

   Domain: jupyterhub.xxxxx.com
   Type:   connection
   Detail: Could not connect to http://jupyterhub.xxxxx.com/.well-
   known/acme-challenge/s7Amlt5iyVBMhANWEizToY7-PdveDtKNgHBVIaIeyP0

   To fix these errors, please make sure that your domain name was
   entered correctly and the DNS A record(s) for that domain
   contain(s) the right IP address. Additionally, please check that
   your computer has a publicly routable IP address and that no
   firewalls are preventing the server from communicating with the
   client. If you're using the webroot plugin, you should also verify
   that you are serving files from the webroot path you provided.
```

**Comment by Analect** [comment]
@jtyberg 
Separate to above, I also tried to get running with self-signed certs, but to no avail.
`openssl req -x509 -nodes -days 365 -newkey rsa:1024 -keyout jupyterhub.key -out jupyterhub.cer`

I still get those same 403 and 500 server errors and this in the jupyterhub logs:
```
jupyterhub |     Traceback (most recent call last):
jupyterhub |       File "/opt/conda/lib/python3.5/site-packages/tornado/web.py", line 1445, in _execute
jupyterhub |         result = yield result
jupyterhub |       File "/srv/jupyterhub/src/oauthenticator/oauthenticator/oauth2.py", line 50, in get
jupyterhub |         username = yield self.authenticator.authenticate(self)
jupyterhub |       File "/srv/jupyterhub/src/oauthenticator/oauthenticator/github.py", line 95, in authenticate
jupyterhub |         access_token = resp_json['access_token']
jupyterhub |     KeyError: 'access_token'
```
If I `docker exec` into the jupyterhub container, I don't see evidence of there being a `/home/jovyan` directory ... as specified in the `README` and `jupyterhub_config.py` file.

I also don't see where that `jupyterhub_config.py` file is copied across from the repo to the image ... or is it enough to pass it as a command in the docker-compose.yml file?

One other thing that might need to get changed in the `Dockerfile.jupyterhub` are the older references to `github.com/jupyter` ... which should now be `github.com/jupyterhub`, I think ...

**Comment by jtyberg** [comment]
@Analect Let's Encrypt can use either port 80 or 443 during cert generation, but you have to specify which one with the `--standalone-supported-challenges` option.  The example script uses port 80

```
docker run --rm -it \
-p 80:80 \
--standalone-supported-challenges http-01 \
...
```

To use 443, do this

```
docker run --rm -it \
-p 443:443 \
--standalone-supported-challenges tls-sni-01 \
...
```

The JupyterHub container runs the proxy and hub processes, not a notebook server, so you won't see `/home/jovyan` directory in the hub container.  

JupyterHub 0.6.0 just released, and several things need to be updated in this deployment, including the dependencies and the base image in `Dockerfile.jupyterhub`.  The base image copies the `jupyterhub_config.py` file into the image during build, but the base image is now called [juptyerhub/jupyterhub-onbuild:0.6](https://github.com/jupyterhub/jupyterhub/tree/master/onbuild).  Right now, the deployment is building `FROM jupyter/jupyterhub:master`, which is incorrect.  This could be what is ailing you.

I've been traveling this week and I'm hopping on a plane now, but I intend to bump everything to JHub 0.6.0 when I return.

**Comment by Analect** [comment]
@jtyberg 
I resolved the letsencrypt problem by making your suggested changes above. Thanks.
Separately, I liased with MinRK at pydata in London on Friday. He was able to help me to add some debugging scripts to this dockerized Jupyterhub. It seems that the root cause of these 403 and 500 responses is to do with the OAuthenticator somehow lower-casing Github usernames. So my Github username starts with a capital letter.  Github doesn't care whether I login in with a lower-cased version.  However, the OAuthenticator somehow lowercases the userlist (whitelist) and then compares with the the 'capitalized' version returned from the Github callback and fails.  I was able to confirm that things work where I use another Github username that happens to be all lower-case.

I think Min is probably best-placed to make the tweak required back in the oauthenticator, so I'll close this out.
cc @minrk 

Thanks. Colum
